gradle/validation/documentation-lint.gradle

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+configure(rootProject) {
+
+  task documentationLint {
+    group 'Verification'
+    description 'Verify all documentation'
+
+    dependsOn 'documentation'
+    dependsOn subprojects.collect { prj ->
+      prj.tasks.matching { task -> task.name == 'checkBrokenLinks'}
+    }
+  }
+
+  // TODO: uncomment this line after fixing all broken links.
+  // (we can't fix the cross-project links until ant build is disabled.)
+  // check.dependsOn documentationLint
+}
+
+configure(subprojects.findAll { it.path in [':lucene', ':solr'] }) {
+  task checkBrokenLinks(type: CheckBrokenLinksTask)
+}
+
+class CheckBrokenLinksTask extends DefaultTask {
+  @Input
+  File docsDir = project.docroot
```

Comment:
(I don't know how to do this without changing field signature) but ideally this should be a lazy provider returning project.docroot. This would make it evaluation-order independent and would trigger input evaluation only if the task actually runs.

Just a thought if you want to get deeper into gradle. ;)

Reply:
Thanks, `project.afterEvaluate()` works for me (though can't figure out the way to lazily evaluate ext properties for the task field).

Reply:
Btw, render-javadoc.gradle (renderSiteJavadoc) also eagarly evaluate the property (I first referred the task and did the same), so it fails if the order is changed in root `build.gradle`. 
https://github.com/apache/lucene-solr/blob/master/gradle/render-javadoc.gradle#L59

Reply:
I wasn't talking about afterEvaluate, actually. I think some properties can be declared as lazy and then assigned a provider - this is different than afterEvaluate. Look here: 
https://docs.gradle.org/current/userguide/lazy_configuration.html

As for order dependencies: afterEvaluate only shifts the problem until later because you still can have dependency ordering problems between afterEvaluate closures... ;) It's not the perfect solution (but of course works reasonably well most of the time).

Reply:
ah... thanks for correction. I misunderstood the meaning of afterEvaluation. I will come back next week.

Reply:
You can have a look at #1477, there I am also using lazy evaluation (see `bindings`) property. It's a MapProvider and the values are lazily evaluated using `project.providers.property{ ... lazy code ... }`

Reply:
You can use a `RegularFileProperty` instead of a File and then use `providers.property{...}` to assign value. Gradle DSL does not change, just when using it from Groovy code or Java code, you have to call `get().getAsFile()` (which does lazy evaluation).
I did not use that for render-javadocs, becuase it was not needed.

Another "simple" way for lazy evaluation (works only with Strings, but its often helpful) is to use GString (double quoted Strings): `GString someProperty = "${->project.docroot}"` (please note the `->`). Whenever `toString()` or `as String` is called, its lazinly evaluated. Without the `->` the value is inserted into the string at parsing time.

I had the same problem with #1477, but I will plan to do this for the URL coming from an extension:

```groovy
  @Input
  def luceneDocUrl = "${->project.luceneDocUrl}"

  @Input
  def solrDocUrl = "${->project.solrDocUrl}"  
```

Reply:
See b7c60e53eb8d7c7c01ccd7c00add2662e503fd08

Reply:
Thanks for the pointers. `project.providers.provider { project.docroot }` perfectly works for me. https://github.com/apache/lucene-solr/pull/1522/commits/ac195d147a3b7d25605ca98034028f9719913782 
(`${->project.docroot}` also works fine; I like using `Provider<File>` signature here.)

On the master branch, `gradlew renderSiteJavadoc` fails with this error when I reverse evaluation order of `gradle/render-javadoc.gradle` and `gradle/documentation/documentation.gradle`.
```
> Could not get unknown property 'docroot' for project ':solr:core' of type org.gradle.api.Project.
```
I think it'd be better doing the same for render-javadoc.gradle just for safety, or add some comments about the ordering on `build.gradle` to prevent someone accidentally changes the file?